### ISSUE TITLE
cov!(cK, k, X1, X2) is inefficient

### ISSUE THREAD

**Issue by maximerischard** [STOR-i/GaussianProcesses.jl#71]
Predictions for exact GPs are too slow. A lot of the blame falls on computing the covariance functions, which is currently much slower than it needs to be.

Here's a demo with a squared exponential kernel. I get a 2x speed-up by using views, and an 8x by doing the distance computations by hand.

```julia
using GaussianProcesses
using PDMats
import GaussianProcesses: cov!, MatF64, VecF64, distance

k = SEIso(0.0, 0.0)
dim = 2
n1,n2 = 1000, 300
X1 = randn(dim, n1)
X2 = randn(dim, n2)
Kx1x2 = Matrix{Float64}(n1, n2)
;

function cov_views!{M1<:MatF64,M2<:MatF64}(cK::MatF64, k::Kernel, X₁::M1, X₂::M2)
    dim, nobs1 = size(X₁)
    nobs2 = size(X₂,2)
    dim == size(X₂,1) || throw(ArgumentError("Input observation matrices must have consistent dimensions"))
    size(cK,1) == nobs1 || throw(ArgumentError(@sprintf("D has %d rows, while X has %d columns (should be same)", 
                                                       size(cK,1), nobs1)))
    size(cK,2) == nobs2 || throw(ArgumentError(@sprintf("D has %d columns, while Y has %d columns (should be same)", 
                                                       size(cK,2), nobs2)))
    @inbounds for i in 1:nobs1, j in 1:nobs2
        x1 = view(X₁, :, i)
        x2 = view(X₂, :, j)
        cK[i,j] = cov(k, x1, x2)
    end

end

function cov_dist!{M1<:MatF64,M2<:MatF64}(cK::MatF64, k::Kernel, X₁::M1, X₂::M2)
    dim, nobs1 = size(X₁)
    nobs2 = size(X₂,2)
    dim == size(X₂,1) || throw(ArgumentError("Input observation matrices must have consistent dimensions"))
    size(cK,1) == nobs1 || throw(ArgumentError(@sprintf("D has %d rows, while X has %d columns (should be same)", 
                                                       size(cK,1), nobs1)))
    size(cK,2) == nobs2 || throw(ArgumentError(@sprintf("D has %d columns, while Y has %d columns (should be same)", 
                                                       size(cK,2), nobs2)))
    @inbounds for i in 1:nobs1, j in 1:nobs2
        s = 0.0
        for d in 1:dim
            s += (X₁[d,i] - X₂[d,j])^2
        end
        R = s
        cK[i,j] = cov(k, R)
    end
end

cov!(Kx1x2, k, X1, X2)
@time for _ in 1:10;cov!(Kx1x2, k, X1, X2);end;
# 0.522648 seconds (13.47 M allocations: 663.254 MiB, 14.12% gc time)

cov_views!(Kx1x2, k, X1, X2)
@time for _ in 1:10;cov_views!(Kx1x2, k, X1, X2);end;
# 0.222245 seconds (6.00 M allocations: 274.658 MiB, 8.92% gc time)

cov_dist!(Kx1x2, k, X1, X2)
@time for _ in 1:10;cov_dist!(Kx1x2, k, X1, X2);end;
# 0.060408 seconds
```

Note how `cov_dist!` doesn't incur any memory allocations: that's how it should be! Using views would be easy enough to implement, but we might need to rethink how we handle distances. `cov_dist!` only works because I knew in advance that the distance metric was squared exponential.

One idea would be to make `Stationary` a parametric abstract type, parametrized by the distance metric type (i.e. SqEuclidean, Euclidean, WeightedSqEuclidean). We would then dispatch the distance computations accordingly. This removes the need to actually instantiate the distance metric types, which I think is the bottle-neck.

**Comment by chris-nemeth** [comment]
This sounds like a good idea. Although for the Weighted distances (used in ARD kernels), the parameters for the metric may need updating periodically. Would this be a problem?

**Comment by maximerischard** [comment]
I've implemented parametric Stationary types in the `faster_predict` branch, and `cov!(Kx1x2, k, X1, X2)` now takes 7 ms and no allocations instead of 50 ms and 66 MiB on my machine.

Currently, this branch is based of the `newoptim` branch, so that I can work on this without installing Klara. I'll wait until that's sorted out before merging. It passes all tests (I've added some extra ones), excluding the MCMC ones.

**Comment by chris-nemeth** [comment]
Wow! That's fast. 

We should be able to make a decision soon on the way forward for the MCMC. In the meantime, I could reintroduce our [HMC implementation](https://github.com/STOR-i/GaussianProcesses.jl/blob/e03d9a52b69b7fc537b2fd5eb27676abdbc6c57f/src/mcmc.jl).

**Comment by maximerischard** [comment]
Merged in bc79eb4
